README.md
# MoneyToNumbers

[![Latest Version on Packagist](https://img.shields.io/packagist/v/logicinception/moneytonumbers.svg?style=flat-square)](https://packagist.org/packages/logicinception/moneytonumbers)
[![Total Downloads](https://img.shields.io/packagist/dt/logicinception/moneytonumbers.svg?style=flat-square)](https://packagist.org/packages/logicinception/moneytonumbers)

Package developed to convert money text on numbers

## Installation

You can install the package via composer:

```bash
composer require logicinception/moneytonumbers
```

## Usage

```php
$textClass = new TextToNumber("dois mil quinhentos e vinte e três reais e dezoito centavos");
$textClass->getNumber(); //output: 2523,18
```

### Testing

```bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

### Security

If you discover any security related issues, please email dev.mengue@gmail.com or itamar.brito@gmail.com instead of using the issue tracker.

## Credits

-   [Wagner Mengue](https://github.com/wagnermengue)
-   [Itamar Brito](https://github.com/Itamar-Brito)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# Changelog

All notable changes to `moneytonumbers` will be documented in this file

## 0.0.1 - 2023-08-29

- initial release


tests/TextToNumberTest.php
<?php

namespace LogicInception\MoneyToNumbers\Tests;

use LogicInception\MoneyToNumbers\TextToNumber;
use PHPUnit\Framework\TestCase;

class TextToNumberTest extends TestCase
{
    /**
     * @dataProvider numbersProvider
     */
    public function test_text_to_number($text, $expected)
    {
        $converted = new TextToNumber($text);
        $this->assertEquals($expected, $converted->getNumber());
    }

    public function numbersProvider(): array
    {
        return [
            ["dois mil quinhentos e vinte e três reais e dezoito centavos", "2523,18"],
            ["sete centavos", "0,07"],
            ["vinte e cinco centavos", "0,25"],
            ["dez reais e dois centavos", "10,02"],
            ["dez reais e um centavo", "10,01"],
            ["cinco mil e cem reais e três centavos", "5100,03"],
//            ["três bilhões e cem mil reais e vinte centavos", "3100000000,00"],
//            ["dois milhões e dez centavos", "2000000,10"],
//            ["quinhentos reais", "500,00"],
//            ["seis milhões e duzentos mil reais e dez centavos", "6200010,00"],
//            ["cento e vinte mil e quinhentos reais", "120500,00"],
//            ["novecentos milhões e setecentos mil reais e oitenta e cinco centavos", "900700000,85"],
//            ["dez mil reais", "10000,00"],
//            ["quarenta e cinco milhões e vinte mil reais e quarenta centavos", "45020000,40"],
//            ["um real", "1,00"],
//            ["vinte mil e trezentos reais", "25300,00"],
//            ["vinte e cinco mil e trezentos reais", "25300,00"],
//            ["dez reais", "10,00"],
        ];
    }
}

src/TextToNumber.php
<?php

namespace LogicInception\MoneyToNumbers;

class TextToNumber
{
    private $simpleNumbers = [1,2,3,4,5,6,7,8,9];

    private $numbers = [
        "um" => 1,
        "dois" => 2,
        "três" => 3,
        "quatro" => 4,
        "cinco" => 5,
        "seis" => 6,
        "sete" => 7,
        "oito" => 8,
        "nove" => 9,
        "dez" => 10,
        "onze" => 11,
        "doze" => 12,
        "treze" => 13,
        "quatorze" => 14,
        "quinze" => 15,
        "dezesseis" => 16,
        "dezessete" => 17,
        "dezoito" => 18,
        "dezenove" => 19,
        "vinte" => 20,
        "trinta" => 30,
        "quarenta"=> 40,
        "cinquenta" => 50,
        "sessenta"=> 60,
        "setenta" => 70,
        "oitenta"=> 80,
        "noventa" => 90,
        "cem"=> 100,
        "cento" => 100,
        "duzentos"=> 200,
        "trezentos" => 300,
        "quatrocentos"=> 400,
        "quinhentos" => 500,
        "seicentos"=> 600,
        "setecentos" => 700,
        "oitocentos"=> 800,
        "novecentos" => 900,
    ];

    private $multiples = [
        "mil" => 1000,
        "milhões" => 1000000,
        "bilhões" => 1000000000
    ];

    private $hasIntegerArgument = true;
    private $hasDecimalArgument = true;
    public function __construct(private string $text)
    {
    }

    public function getNumber(): string
    {

        $textSanitized = str_replace(' e ', ' ', $this->text);
        $explodedMoney = [];
        try {
            $explodedMoney = $this->explodeTextMoney($textSanitized);
        } catch (\Exception $exception) {
            $this->hasIntegerArgument = false;
        }
        print_r($explodedMoney);
        $integerArgument = "0";
        if ($this->hasIntegerArgument) {
            $arguments = explode(' ', $explodedMoney[0]);
            $integerArgument = $this->integerArgument($arguments);
        }
        try {
            $stringCents = empty($explodedMoney) ? $textSanitized : $explodedMoney[1];
            $decimalArguments = $this->explodeCents($stringCents);
            $decimalArgument = $this->decimalArgument($decimalArguments);
        } catch (\Exception $exception) {
            $decimalArgument = "00";
        }


        return "$integerArgument,$decimalArgument";
    }

    private function explodeTextMoney(string $textSanitized): array
    {
        if(str_contains($textSanitized, 'real')) {
            $explodedArray = explode('real', $textSanitized);
            return $this->sanitizeArray($explodedArray);
        }
        if(str_contains($textSanitized, 'reais')) {
            $explodedArray = explode('reais', $textSanitized);
            return $this->sanitizeArray($explodedArray);
        }
        throw new \Exception('Without integer value');
    }

    private function explodeCents(string $decimalArgument): array
    {
        if(str_contains($decimalArgument, 'centavos')) {
            $sanitizeCents = str_replace('centavos', '', $decimalArgument);
            $explodedArray = explode(' ', $sanitizeCents);
            return $this->sanitizeArray($explodedArray);
        }
        $sanitizeCents = str_replace('centavo', '', $decimalArgument);
        $explodedArray = explode(' ', $sanitizeCents);
        return $this->sanitizeArray($explodedArray);
    }

    private function sanitizeArray(array $array)
    {
        return array_map('trim', array_filter($array));
    }

    private function integerArgument($arguments)
    {
        $output = [];
        foreach ($arguments as $key => $number) {
            if ($this->isMultiple($number)) {
                $output[$key-1] = $output[$key-1] * $this->multiples[$number];
                continue;
            }
            $output[] = $this->numbers[$number];
        }
        return array_sum($output);
    }

    private function isMultiple($number)
    {
        return array_key_exists($number, $this->multiples);
    }

    private function decimalArgument($decimalArguments)
    {
        $output = [];
        foreach ($decimalArguments as $number) {
            $output[] = $this->numbers[$number];
        }
        $decimalOutput = array_sum($output);

        if ($decimalOutput == "0") {
            $decimalOutput = "00";
        }

        if (in_array($decimalOutput, $this->simpleNumbers)) {
            $decimalOutput = str_pad($decimalOutput, 2, '0', STR_PAD_LEFT);
        }
        return $decimalOutput;
    }
}